runners/flink/1.6/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -0,0 +1,959 @@
+   * with a timestamp {@code < T} in the future. A Flink watermark {@code T} says there will not be
+   * any elements with a timestamp {@code <= T} in the future. We correct this by subtracting {@code
+   * 1} from a Beam watermark before passing to any relevant Flink runtime components.
+   */
+  private static long toFlinkRuntimeWatermark(long beamWatermark) {
+    return beamWatermark - 1;
+  }
+
+  /**
+   * Emits all pushed-back data. This should be used once we know that there will not be any future
+   * side input, i.e. that there is no point in waiting.
+   */
+  private void emitAllPushedBackData() throws Exception {
+
+    Iterator<WindowedValue<InputT>> it = pushedBackElementsHandler.getElements().iterator();
+
+    while (it.hasNext()) {
+      WindowedValue<InputT> element = it.next();
+      // we need to set the correct key in case the operator is
+      // a (keyed) window operator
+      setKeyContextElement1(new StreamRecord<>(element));
+
+      doFnRunner.processElement(element);
+    }
+
+    pushedBackElementsHandler.clear();
+
+    setPushedBackWatermark(Long.MAX_VALUE);
+  }
+
+  /**
+   * Check whether invoke startBundle, if it is, need to output elements that were buffered as part
+   * of finishing a bundle in snapshot() first.
+   *
+   * <p>In order to avoid having {@link DoFnRunner#processElement(WindowedValue)} or {@link
+   * DoFnRunner#onTimer(String, BoundedWindow, Instant, TimeDomain)} not between StartBundle and
+   * FinishBundle, this method needs to be called in each processElement and each processWatermark
+   * and onProcessingTime. Do not need to call in onEventTime, because it has been guaranteed in the
+   * processWatermark.
+   */
+  private void checkInvokeStartBundle() {
+    if (bundleStarted.compareAndSet(false, true)) {
+      pushbackDoFnRunner.startBundle();
+    }
+  }
+
+  /** Check whether invoke finishBundle by elements count. Called in processElement. */
+  private void checkInvokeFinishBundleByCount() {
+    elementCount++;
+    if (elementCount >= maxBundleSize) {
+      invokeFinishBundle();
+    }
+  }
+
+  /** Check whether invoke finishBundle by timeout. */
+  private void checkInvokeFinishBundleByTime() {
+    long now = getProcessingTimeService().getCurrentProcessingTime();
+    if (now - lastFinishBundleTime >= maxBundleTimeMills) {
+      invokeFinishBundle();
+    }
+  }
+
+  protected final void invokeFinishBundle() {
+    if (bundleStarted.compareAndSet(true, false)) {
+      pushbackDoFnRunner.finishBundle();
+      elementCount = 0L;
+      lastFinishBundleTime = getProcessingTimeService().getCurrentProcessingTime();
+    }
+  }
+
+  @Override
+  public void prepareSnapshotPreBarrier(long checkpointId) {
+    // Finish the current bundle before the snapshot barrier is sent downstream
+    // This give us a clean state before taking the actual snapshot
+    invokeFinishBundle();
```

Comment:
I should highlight that this is the new method used to finish the bundle before the snapshot barrier is emitted and `snapshot(..)` gets called.